src/Featurizers/TfidfVectorizerFeaturizer.h

```diff
@@ -0,0 +1,523 @@
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    IndexMap const                           Labels;
+    FrequencyMap const                       DocumentFreq;
+    std::uint32_t const                      TotalNumsDocuments;
+    //Binary: If True, all non zero counts are set to 1. This is useful for discrete probabilistic models that model binary events rather than integer counts.
+    bool const                               Binary;
+    std::string const                        Norm;
+    bool const                               UseIdf;
+    bool const                               SmoothIdf;
+    bool const                               SublinearTf;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    TfidfVectorizerTransformer(
+        IndexMap labels,
+        FrequencyMap docufreq,
+        std::uint32_t totalnumdocus,
+        bool binary,
+        std::string norm,
+        bool use_idf,
+        bool smooth_idf,
+        bool sublinear_tf
+    );
+    TfidfVectorizerTransformer(Archive &ar);
+
+    ~TfidfVectorizerTransformer(void) override = default;
+
+    void save(Archive &ar) const override;
+
+    bool operator==(TfidfVectorizerTransformer const &other) const;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC has problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        //termfrequency for specific document
+        MapWithIterRange documentTermFrequency;
+
+        //todo: will use vector<functor> after string header file is done
+        Strings::Parse<std::string::const_iterator>(
+            input,
+            [](char c) {return std::isspace(c);},
+            [&documentTermFrequency] (std::string::const_iterator iterStart, std::string::const_iterator iterEnd) {
+                MapWithIterRange::iterator docuTermFreqIter(documentTermFrequency.find(std::make_tuple(iterStart, iterEnd)));
+                if (docuTermFreqIter != documentTermFrequency.end()) {
+                    ++docuTermFreqIter->second;
+                } else {
+                    documentTermFrequency.insert(std::make_pair(std::make_tuple(iterStart, iterEnd), 1));
+                }
+            }
+        );
+
+        std::float_t normVal = 0.0f;
+        std::vector<std::tuple<std::uint32_t, std::float_t>> results;
+        for (auto const & pair : documentTermFrequency) {
```

Comment:
>pair [](start = 26, length = 4)

Consider a different name here - pair describes the type (I think) rather than what the data actually is.